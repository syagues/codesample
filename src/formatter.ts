import Prism from 'prismjs'

export const setup = () => {
  Prism.hooks.add('before-highlight', (env) => {
    addLineNumbers(env.grammar);
  });
}

export const format = (content: string, language: string) => {
  return Prism.highlight(content, Prism.languages[language], language);
}

const addLineNumbers = (grammar) => {
  console.log(grammar);
  
}