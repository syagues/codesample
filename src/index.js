import Prism from 'prismjs'

const code = `const data = 'example'`

const html = Prism.highlight(code, Prism.languages.javascript, 'javascript');