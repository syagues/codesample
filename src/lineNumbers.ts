import vsctm from 'vscode-textmate';
import oniguruma from 'vscode-oniguruma';
// @ts-ignore
import wasmBin from 'vscode-oniguruma/release/onig.wasm';

const generateCode = ($element: Element) => {
    const $pre = document.createElement('pre')
    const $code = document.createElement('code')
    $pre.appendChild($code)
    $element.appendChild($pre)

    return $code
}

const vscodeOnigurumaLib = oniguruma.loadWASM(wasmBin).then(() => {
    return {
        createOnigScanner(patterns) { return new oniguruma.OnigScanner(patterns); },
        createOnigString(s) { return new oniguruma.OnigString(s); }
    };
});

// Create a registry that can create a grammar from a scope name.
const registry = new vsctm.Registry({
    onigLib: vscodeOnigurumaLib,
    loadGrammar: (scopeName) => {
        console.log(scopeName);
        
        if (scopeName === 'source.js') {
            return fetch('https://raw.githubusercontent.com/textmate/javascript.tmbundle/master/Syntaxes/JavaScript.plist')
              .then(response => response.text())
              .then(content => vsctm.parseRawGrammar(content));
        }
        console.log(`Unknown scope name: ${scopeName}`);
        return Promise.resolve(null);
    }
});

// Load the JavaScript grammar and any other grammars included by it async.
const execute = ($element: Element) => {
    registry.loadGrammar('source.js').then(grammar => {
        console.log(grammar);
        
        if (!grammar) return
        const text = [
            `// There is a comment`,
            `function sayHello(name) {`,
            `\treturn "Hello, " + name;`,
            `}`
        ];
        let ruleStack = vsctm.INITIAL;
        const $code = generateCode($element)
        for (let i = 0; i < text.length; i++) {
            const line = text[i];
            const lineTokens = grammar.tokenizeLine(line, ruleStack);
            console.log(`\nTokenizing line: ${line}`);
            const $line = document.createElement('div')
            $line.classList.add('line')
            for (let j = 0; j < lineTokens.tokens.length; j++) {
                const token = lineTokens.tokens[j];
                console.log(` - token from ${token.startIndex} to ${token.endIndex} ` +
                  `(${line.substring(token.startIndex, token.endIndex)}) ` +
                  `with scopes ${token.scopes.join(', ')}`
                );
                const $token = document.createElement('span')
                $token.classList.add('token')
                $token.innerHTML = line.substring(token.startIndex, token.endIndex)
                $line.appendChild($token)
            }
            ruleStack = lineTokens.ruleStack;
            $code.appendChild($line)
        }
    });
}
export default execute

/* OUTPUT:

Unknown scope name: source.js.regexp

Tokenizing line: function sayHello(name) {
 - token from 0 to 8 (function) with scopes source.js, meta.function.js, storage.type.function.js
 - token from 8 to 9 ( ) with scopes source.js, meta.function.js
 - token from 9 to 17 (sayHello) with scopes source.js, meta.function.js, entity.name.function.js
 - token from 17 to 18 (() with scopes source.js, meta.function.js, punctuation.definition.parameters.begin.js
 - token from 18 to 22 (name) with scopes source.js, meta.function.js, variable.parameter.function.js
 - token from 22 to 23 ()) with scopes source.js, meta.function.js, punctuation.definition.parameters.end.js
 - token from 23 to 24 ( ) with scopes source.js
 - token from 24 to 25 ({) with scopes source.js, punctuation.section.scope.begin.js

Tokenizing line:        return "Hello, " + name;
 - token from 0 to 1 (  ) with scopes source.js
 - token from 1 to 7 (return) with scopes source.js, keyword.control.js
 - token from 7 to 8 ( ) with scopes source.js
 - token from 8 to 9 (") with scopes source.js, string.quoted.double.js, punctuation.definition.string.begin.js
 - token from 9 to 16 (Hello, ) with scopes source.js, string.quoted.double.js
 - token from 16 to 17 (") with scopes source.js, string.quoted.double.js, punctuation.definition.string.end.js
 - token from 17 to 18 ( ) with scopes source.js
 - token from 18 to 19 (+) with scopes source.js, keyword.operator.arithmetic.js
 - token from 19 to 20 ( ) with scopes source.js
 - token from 20 to 24 (name) with scopes source.js, support.constant.dom.js
 - token from 24 to 25 (;) with scopes source.js, punctuation.terminator.statement.js

Tokenizing line: }
 - token from 0 to 1 (}) with scopes source.js, punctuation.section.scope.end.js

*/
