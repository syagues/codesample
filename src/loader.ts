import defTheme from './themes/solarized.css?inline'

export const loadTheme = (theme: string) => {
  if (theme === 'nord') {
    const $style = document.createElement('style')
    $style.id = 'csTheme'
    $style.innerHTML = defTheme
    document.head.appendChild($style)
  }
}
