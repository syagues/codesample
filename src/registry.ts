import vsctm from 'vscode-textmate';
import oniguruma from 'vscode-oniguruma';
// @ts-ignore
import wasmBin from 'vscode-oniguruma/release/onig.wasm';

const vscodeOnigurumaLib = oniguruma.loadWASM(wasmBin).then(() => {
  return {
      createOnigScanner(patterns) { return new oniguruma.OnigScanner(patterns); },
      createOnigString(s) { return new oniguruma.OnigString(s); }
  };
});

// Create a registry that can create a grammar from a scope name.
export default new vsctm.Registry({
  onigLib: vscodeOnigurumaLib,
  loadGrammar: (scopeName) => {
      console.log(scopeName);
      
      if (scopeName === 'source.js') {
          return fetch('https://raw.githubusercontent.com/textmate/javascript.tmbundle/master/Syntaxes/JavaScript.plist')
            .then(response => response.text())
            .then(content => vsctm.parseRawGrammar(content));
      }
      console.log(`Unknown scope name: ${scopeName}`);
      return Promise.resolve(null);
  }
});