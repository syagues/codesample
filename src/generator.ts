import { format } from './formatter'

export const generateCodeSamples = async (
  $element: Element,
  samples: samples
) => {
  const $wrapper = document.createElement('div')
  $wrapper.classList.add('cs')
  for (const sampleKey in samples) {
    const $sample: Element = document.createElement('div')
    $sample.classList.add('sample')
    $sample.classList.add(sampleKey)
    const sample = samples[sampleKey]
    format($sample, sample.content, sample.language)
    $wrapper.appendChild($sample)
  }
  $element.appendChild($wrapper)
}
