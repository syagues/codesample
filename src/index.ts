import { generateCodeSamples } from './generator'
import { loadTheme } from './loader'

export class CodeSample {
  theme: string
  samples: samples

  constructor(options) {
    this.theme = options.theme || 'nord'
    this.samples = options.samples || {}

    loadTheme(this.theme)
  }

  /**
   * Attach the generated CodeSample to the given
   * DOM selector
   * @param selector String
   */
  async attachToElement(selector: string) {
    const $element: Element | null = document.querySelector(selector)
    if (!$element) throw new Error(`selector ${selector} is not in the DOM`)

    await generateCodeSamples($element, this.samples)
  }

  /**
   * Gets the generated CodeSample as a String
   * @returns String
   */
  async getHtml() {
    return ''
  }

  /**
   * Add a new sample to the samples object
   * @param sampleKey string
   * @param options object
   */
  addSample(sampleKey: string, options: sample) {
    if (this.samples[sampleKey])
      throw new Error(`Key [${sampleKey}] already existing`)
    if (!options.language) throw new Error('Language not defined')
    if (!options.content) throw new Error('Content not defined')

    this.samples[sampleKey] = options
  }

  /**
   * Removes the given sample from samples object
   * @param sampleKey string
   */
  removeSample(sampleKey: string) {
    if (!this.samples[sampleKey])
      throw new Error(`Key [${sampleKey}] not existing`)

    delete this.samples[sampleKey]
  }

  /**
   * Hichlights the indicated block of code using
   * classes
   * @param sampleKey string
   * @param location Array<number>
   */
  highlightBlock(sampleKey: string, location: Array<number>) {
    return Promise.resolve()
  }

  /**
   * Scrolls the view to the given location.
   * @param sampleKey string
   * @param location Array<number>
   */
  goToBlock(sampleKey: string, location: Array<number>) {
    return Promise.resolve()
  }

  /**
   * Activates the sample with the given key.
   * @param sampleKey string
   */
  navigateToSample(sampleKey: string) {
    return Promise.resolve()
  }
}
