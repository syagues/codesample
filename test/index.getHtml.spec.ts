import { expect, test } from 'vitest'
import { CodeSample } from '../src/index'

test('getHtml :: get 1 sample html', () => {
  const options = {
    theme: 'nord',
    samples: {
      // Id
      sample: {
        language: 'javascript',
        fileName: 'sample.js', // optional
        content: `const foo = 'bar'`
      }
    }
  }
  
  const codeSample = new CodeSample(options)
  const html = codeSample.getHtml()
  expect(html).toMatchSnapshot()
})