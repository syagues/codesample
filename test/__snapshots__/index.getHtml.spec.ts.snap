// Vitest Snapshot v1

exports[`getHtml :: get 1 sample html 1`] = `
"<div class=\\"cs\\"><div class=\\"sample sample\\">
      <pre><code><div class=\\"line\\"><span class=\\"line-number\\">1</span></span><span class=\\"hljs-keyword\\">const</span> foo = <span class=\\"hljs-string\\">&#x27;bar&#x27;</span></div></code></pre>
    </div></div>"
`;
